docs/source/basics/custom_components.rst

```diff
@@ -0,0 +1,171 @@
+    # Or launch directly from the project directory with a relative path
+    tune run full_finetune_single_device --config config/qwen_config.yaml
+
+For a more detailed discussion on downloading models and modifying library configs, see :ref:`finetune_llama_label`.
+
+Launching custom recipes
+------------------------
+torchtune's built-in recipes provide starting points for your fine-tuning workflows, but you can write your own training loop
+with customized logic for your use case and launch training with ``tune run``. Similar to modifying library configs, you can
+also copy one of our recipes as a starting point and modify, or write one completely from scratch. Note that for launching
+custom recipes, you must specify the file extension.
+
+.. code-block:: bash
+
+    mkdir ~/my_project/recipes
+    # Show all the default recipes
+    tune ls
+    # This makes a copy of the full finetune single device recipe locally
+    tune cp full_finetune_single_device ~/my_project/recipes/single_device.py
+    # Launch custom recipe with custom config from project directory
+    tune run recipes/single_device.py --config config/qwen_config.yaml
+
+If you are writing a new recipe from scratch, we recommend following the Python convention of defining a ``main()`` function
+in your script and decorating it with the :func:`~torchtune.config.parse` decorator. This will enable you to launch the recipe
+with ``tune run`` and pass in a yaml file for the ``--config`` argument.
+
+.. code-block:: python
+
+    from torchtune import config
+    from omegaconf import DictConfig
+
+    @config.parse
+    def main(cfg: DictConfig):
+        # Add all your recipe logic here, access config fields as attributes
+
+    if __name__ == "__main__":
+        # Config will be parsed from CLI, don't need to pass in here
+        main()
+
+Launching with custom components
+--------------------------------
+torchtune supports full experimentation with custom models, datasets, optimizers, or any fine-tuning component. You can define
+these locally in your repo and use them in your recipes and configs that you can launch with ``tune run``.
+
+We recommend following the "builder" pattern when making your components. This means creating "builder" functions that set up
+the classes you need with a few high level parameters that can be modified easily from the config. For example, we can define custom
+model and dataset builders in our project directory:
+
+.. code-block:: python
+
+    #
+    # In models/custom_decoder.py
+    #
+    class CustomTransformerDecoder(nn.Module):
+        # A custom architecture not present in torchtune
+
+    # Builder function for the custom model
+    def custom_model(num_layers: int, classification_head: bool = False):
+        # Any setup for defining the class
+        ...
+        # Return the module you want to train
+        return CustomTransformerDecoder(...)
+
```

Comment:
```suggestion

This allows us to expose our custom model in a config friendly manner - rather than having to re-define every argument needed to construct our custom model, we only expose the arguments which we care about modifying. This is how we implement our models in torchtune - see :func:`torchtune.models.llama3_2_vision.llama3_2_vision_11b` as an example!

```
just a suggestion to help be super obvious why we reccomend doing it this way, I'm not sure if everyone would clock this design decision immediately